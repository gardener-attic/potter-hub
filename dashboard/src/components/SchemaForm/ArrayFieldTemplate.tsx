import * as React from "react";
import { ArrayFieldTemplateProps } from "react-jsonschema-form";

// adapted from https://github.com/mozilla-services/react-jsonschema-form/blob/master/playground/samples/customArray.js
const ArrayFieldTemplate: React.SFC<ArrayFieldTemplateProps> = props => {
  const { className, title, items, canAdd, onAddClick } = props;
  return (
    <div className={`${className} margin-b-normal`}>
      <label>{title}</label>
      {items &&
        items.map(element => (
          <div key={element.index}>
            <div className="margin-b-normal">{element.children}</div>
            {element.hasMoveDown && (
              <button
                type="button"
                className="button button-small"
                onClick={element.onReorderClick(element.index, element.index + 1)}
              >
                Down
              </button>
            )}
            {element.hasMoveUp && (
              <button
                type="button"
                className="button button-small"
                onClick={element.onReorderClick(element.index, element.index - 1)}
              >
                Up
              </button>
            )}
            <button
              type="button"
              className="button button-danger button-small"
              onClick={element.onDropIndexClick(element.index)}
            >
              Delete
            </button>
            <hr />
          </div>
        ))}

      {canAdd && (
        <div className="row">
          <button className="button button-primary button-small" type="button" onClick={onAddClick}>
            Add Item
          </button>
        </div>
      )}
    </div>
  );
};

export default ArrayFieldTemplate;
