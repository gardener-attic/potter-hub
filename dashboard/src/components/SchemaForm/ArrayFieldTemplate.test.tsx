import { shallow } from "enzyme";
import * as React from "react";
import { ArrayFieldTemplateProps } from "react-jsonschema-form";

import ArrayFieldTemplate from "./ArrayFieldTemplate";

it("renders a label for the title", () => {
  const wrapper = shallow(<ArrayFieldTemplate {...({} as ArrayFieldTemplateProps)} title="Test" />);
  expect(wrapper.find("label").text()).toBe("Test");
});

it("renders each element in the array", () => {
  const items = [
    {
      children: <div className="test1" />,
      hasMoveDown: true,
      hasMoveUp: true,
      index: 0,
      onDropIndexClick: (i: number) => {
        jest.fn();
      },
      onReorderClick: (i: number, j: number) => {
        jest.fn();
      },
    } as ArrayFieldTemplateProps["items"][0],
    {
      children: <div className="test2" />,
      hasMoveDown: true,
      hasMoveUp: true,
      index: 1,
      onDropIndexClick: (i: number) => {
        jest.fn();
      },
      onReorderClick: (i: number, j: number) => {
        jest.fn();
      },
    },
  ] as ArrayFieldTemplateProps["items"];
  const wrapper = shallow(
    <ArrayFieldTemplate {...({} as ArrayFieldTemplateProps)} items={items} />,
  );
  expect(wrapper.find(".test1").exists()).toBe(true);
  expect(wrapper.find(".test2").exists()).toBe(true);
  expect(wrapper).toMatchSnapshot();
});

it("renders the add item button if enabled", () => {
  const onAddClick = jest.fn();
  const wrapper = shallow(
    <ArrayFieldTemplate
      {...({} as ArrayFieldTemplateProps)}
      canAdd={true}
      onAddClick={onAddClick}
    />,
  );
  const button = wrapper.find("button.button-primary");
  expect(button.exists()).toBe(true);
  button.simulate("click");
  expect(onAddClick).toBeCalled();
});
