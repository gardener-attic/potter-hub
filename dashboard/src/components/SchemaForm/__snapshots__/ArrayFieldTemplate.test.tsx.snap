// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders each element in the array 1`] = `
<div
  className="undefined margin-b-normal"
>
  <label />
  <div
    key="0"
  >
    <div
      className="margin-b-normal"
    >
      <div
        className="test1"
      />
    </div>
    <button
      className="button button-small"
      type="button"
    >
      Down
    </button>
    <button
      className="button button-small"
      type="button"
    >
      Up
    </button>
    <button
      className="button button-danger button-small"
      type="button"
    >
      Delete
    </button>
    <hr />
  </div>
  <div
    key="1"
  >
    <div
      className="margin-b-normal"
    >
      <div
        className="test2"
      />
    </div>
    <button
      className="button button-small"
      type="button"
    >
      Down
    </button>
    <button
      className="button button-small"
      type="button"
    >
      Up
    </button>
    <button
      className="button button-danger button-small"
      type="button"
    >
      Delete
    </button>
    <hr />
  </div>
</div>
`;
